plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/SpringChangedHub.java

```diff
@@ -0,0 +1,267 @@
+        }
+    }
+
+    public static void addChangedProperty(URL property) {
+        for (SpringChangedHub springChangedHub : springChangedHubs.values()) {
+            springChangedHub.addProperty(property);
+        }
+    }
+
+    void addClass(Class clazz) {
+        if (status.compareAndSet(INIT, CHANGING) || status.intValue() == CHANGING) {
+            current.addClass(clazz);
+        } else if (status.intValue() == RELOADED) {
+            next.addClass(clazz);
+        }
+    }
+
+    void addProperty(URL property) {
+        if (status.compareAndSet(INIT, CHANGING) || status.intValue() == CHANGING) {
+            current.addProperty(property);
+        } else if (status.intValue() == RELOADED) {
+            next.addProperty(property);
+        }
+    }
+
+    void addXml(URL xml) {
+        if (status.compareAndSet(INIT, CHANGING) || status.intValue() == CHANGING) {
+            current.addXml(xml);
+        } else if (status.intValue() == RELOADED) {
+            next.addXml(xml);
+        }
+    }
+
+    void addNewBean(BeanDefinitionRegistry registry, BeanDefinitionHolder beanDefinitionHolder) {
+        if (status.compareAndSet(INIT, CHANGING) || status.intValue() == CHANGING) {
+            current.addScanNewBean(registry, beanDefinitionHolder);
+        } else if (status.intValue() == RELOADED) {
+            next.addScanNewBean(registry, beanDefinitionHolder);
+        }
+    }
+
+    private void reload() {
+        // delay three times
+        if (waitTimes.get() >= maxWaitTimes) {
+            waitTimes.incrementAndGet();
+            if (pause.get()) {
+                LOGGER.trace("spring reload pause: {}", ObjectUtils.identityToString(current.beanFactory));
+                return;
+            }
+            if (status.compareAndSet(CHANGING, RELOADED)) {
+                // relead
+                try {
+                    current.reload();
+                } catch (Exception e) {
+                    LOGGER.error("reload spring failed: {}", e, ObjectUtils.identityToString(current.beanFactory));
+                }
+                finishReload();
+                waitTimes.set(0);
+            }
+        } else if (status.get() == CHANGING) {
+            if (waitTimes.get() == 0) {
+                int restTime = maxWaitTimes * DEFAULT_DELAY_PERIOD / 1000;
+                LOGGER.info("waiting to start reload '{}', it will start after {}s", ObjectUtils.identityToString(beanFactory()), restTime);
+            }
+            waitTimes.incrementAndGet();
+        }
+    }
+
+    public static void collectPlaceholderProperties(ConfigurableListableBeanFactory configurableListableBeanFactory) {
+        if (!(configurableListableBeanFactory instanceof DefaultListableBeanFactory)) {
+            return;
+        }
+        getInstance((DefaultListableBeanFactory) configurableListableBeanFactory).doCollectPlaceHolderProperties();
+    }
+
+    private void doCollectPlaceHolderProperties() {
```

Comment:
I think all placeholder relevant logics should be hidden in SpringReload instead of this class.